org.adempiere.base/src/org/adempiere/base/AnnotationBasedModelFactory.java

```diff
@@ -0,0 +1,65 @@
+package org.adempiere.base;
+
+import org.atteo.classindex.ClassIndex;
+import org.osgi.framework.Bundle;
+import org.osgi.framework.wiring.BundleWiring;
+import org.osgi.service.component.ComponentContext;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.ServiceScope;
+
+/**
+ * Translates table names into model classes having the {@link Model} annotation. Relies on
+ * {@link DefaultModelFactory} for everything else.
+ * This factory is designed to have a service rank higher than {@link DefaultModelFactory}, as class
+ * discovery using SPI is preferred over reflection-based methods.
+ * @author Saulo Gil
+ */
+@Component(scope = ServiceScope.BUNDLE)
+public class AnnotationBasedModelFactory extends DefaultModelFactory implements IModelFactory
+{
+
+	private Bundle usingBundle;
+
+	@Activate
+	void activate(ComponentContext context)
+	{
+		this.usingBundle = context.getUsingBundle();
+	}
+
+	@Override
+	public Class<?> getClass(String tableName) 
+	{
+		// search first into DefaultModelFactory's cache
+		Class<?> candidate = s_classCache.get(tableName);
+		if (candidate != null)
+		{
+			// Object.class indicate no generated PO class for tableName
+			if (candidate.equals(Object.class))
+				return null;
+			else
+				return candidate;
+		}
+
+		// scan annotations
+		BundleWiring wiring = usingBundle.adapt(BundleWiring.class);
+		for(Class<?> clazz : ClassIndex.getAnnotated(Model.class, wiring.getClassLoader()))
+		{
+			Model ma = clazz.getAnnotation(Model.class);
+			if(ma.table().equalsIgnoreCase(tableName)) 
+			{
+				candidate = clazz;
+				// M* classes are preferred over X* classes
+				if(!ma.intermediate())
+					break;
+			}
+		}
+
+		if(candidate!=null)
+			// inject association into DefaultModelFactory's cache
+			s_classCache.put(tableName, candidate);
+
+		return candidate;
+	}
+
+}
```

Comment:
I don't see a need to extend DefaultModelFactory here. Since we are not removing DefaultModelFactory, we don't have to do an extend here - if AnnotationBasedModelFactory return null, the DefaultModelFactory will just do its part as it is now. Also, if we drop DefaultModelFactory in future, this is just unnecessary legacy baggage.

Reply:
The main reason behind extending ```DefaultModelFactory``` is to avoid rewriting ```org.adempiere.base.DefaultModelFactory.getPO(String, int, String)```. It might not be ideal to have every other model factory implement ```PO``` instantiation via reflection on it's own, duplicating the same code each time (e.g.: https://github.com/ingeint/idempiere-plugin-scaffold/blob/master/template/plugin.symbolic.name/src/plugin.root/base/CustomModelFactory.java#L78 ).
If there's an intention to deprecate ```DefaultModelFactory``` I'd suggest to move this kind of model instantiation somewhere else, so that it can be reused by model factories (especially coming from plugins). Maybe we could also cache instances of these standard constructors and speed up model class instantiation a little bit.

Reply:
alright, that's a good point. I guess we can do the refactoring and performance optimization later (the straightforward approach is to extract that to an abstractmodelfactory class and have the class cache as instance variable instead. The class cache doesn't have to be static since osgi service is already a singleton).

Reply:
OK I'll try that approach on a separate PR.